src/isomorphic/hooks/ReactComponentTreeHook.js

```diff
@@ -94,88 +101,93 @@ function describeID(id) {
-        );
-        invariant(
-          nextChild.isMounted,
-          'Expected onMountComponent() to fire for the child ' +
-          'before its parent includes it in onSetChildren().'
-        );
-        if (nextChild.parentID == null) {
-          nextChild.parentID = id;
-          // TODO: This shouldn't be necessary but mounting a new root during in
-          // componentWillMount currently causes not-yet-mounted components to
-          // be purged from our tree data so their parent ID is missing.
-        }
-        invariant(
-          nextChild.parentID === id,
-          'Expected onSetParent() and onSetChildren() to be consistent (%s ' +
-          'has parents %s and %s).',
-          nextChildID,
-          nextChild.parentID,
-          id
-        );
-      });
-    });
-  },
-
-  onSetOwner(id, ownerID) {
-    update(id, item => item.ownerID = ownerID);
+    var item = get(id);
+    item.childIDs = nextChildIDs;
+
+    for (var i = 0; i < nextChildIDs.length; i++) {
+      var nextChildID = nextChildIDs[i];
+      var nextChild = get(nextChildID);
+      invariant(
+        nextChild,
+        'Expected hook events to fire for the child ' +
+        'before its parent includes it in onSetChildren().'
+      );
+      invariant(
+        nextChild.childIDs != null ||
+        typeof nextChild.element !== 'object' ||
+        nextChild.element == null,
+        'Expected onSetChildren() to fire for a container child ' +
+        'before its parent includes it in onSetChildren().'
+      );
+      invariant(
+        nextChild.isMounted,
+        'Expected onMountComponent() to fire for the child ' +
+        'before its parent includes it in onSetChildren().'
+      );
+      if (nextChild.parentID == null) {
+        nextChild.parentID = id;
+        // TODO: This shouldn't be necessary but mounting a new root during in
+        // componentWillMount currently causes not-yet-mounted components to
+        // be purged from our tree data so their parent ID is missing.
+      }
+      invariant(
+        nextChild.parentID === id,
+        'Expected onSetParent() and onSetChildren() to be consistent (%s ' +
+        'has parents %s and %s).',
+        nextChildID,
+        nextChild.parentID,
+        id
+      );
+    }
   },
 
   onSetParent(id, parentID) {
-    update(id, item => item.parentID = parentID);
+    var item = get(id);
+    item.parentID = parentID;
   },
 
-  onSetText(id, text) {
-    update(id, item => item.text = text);
+  onInstantiateComponent(id, element) {
+    create(id, element);
   },
 
   onBeforeMountComponent(id, element) {
-    update(id, item => item.element = element);
+    var item = get(id);
+    item.element = element;
   },
 
   onBeforeUpdateComponent(id, element) {
-    update(id, item => item.element = element);
+    var item = get(id);
+    if (!item || !item.isMounted) {
+      // We may end up here as a result of setState() in componentWillUnmount().
```

Comment:
Hmm. I didn't think we should hit this on unmounted components. setState should bail out before we get to this point. Is there a test that fails without this?